src/components/focus-trap/focus-trap.ts

```diff
@@ -0,0 +1,198 @@
+  }
+
+  public focusLastElement() {
+    this.focusableElements.at(-1)?.focus();
+  }
+
+  protected override render() {
+    const tabStop = !this.focused || this.disabled ? -1 : 0;
+
+    return html`
+      <div
+        id="start"
+        tabindex=${tabStop}
+        @focus=${this.disabled ? nothing : this.focusLastElement}
+      ></div>
+      <slot></slot>
+      <div
+        id="end"
+        tabindex=${tabStop}
+        @focus=${this.disabled ? nothing : this.focusFirstElement}
+      ></div>
+    `;
+  }
+}
+
+declare global {
+  interface HTMLElementTagNameMap {
+    'igc-focus-trap': IgcFocusTrapComponent;
+  }
+}
+
+const defaultSelectors = [
+  '[tabindex]',
+  'a[href]',
+  'button',
+  'input',
+  'select',
+  'textarea',
+];
+
+/** Returns whether the element is hidden. */
+function isHidden(node: HTMLElement) {
+  return (
+    node.hasAttribute('hidden') ||
+    node.hasAttribute('inert') ||
+    (node.hasAttribute('aria-hidden') &&
+      node.getAttribute('aria-hidden') !== 'false')
+  );
+}
+
+/** Returns whether the element is disabled. */
+function isDisabled(node: HTMLElement) {
+  return node.hasAttribute('disabled') || node.hasAttribute('inert');
+}
+
+/**
+ * Returns whether the element can be focused.
+ */
+function isFocusable(node: HTMLElement) {
+  if (node.tabIndex === -1 || isHidden(node) || isDisabled(node)) {
+    return false;
+  }
+
+  return defaultSelectors.some((selector) => node.matches(selector));
+}
+
+/**
+ * Filter function for the tree walker instance skipping over nodes and their children
+ * if the `node` is hidden/disabled or it was already visited and resides in `cache`.
+ */
+function shouldSkipElements(node: Node, cache?: WeakSet<HTMLElement>) {
+  const element = node as HTMLElement;
+
+  return isHidden(element) || isDisabled(element) || cache?.has(element)
+    ? NodeFilter.FILTER_REJECT
+    : NodeFilter.FILTER_ACCEPT;
+}
+
+/** Returns the slotted elements and the parent element containing the slot */
+function getSlottedElements(node: HTMLElement) {
+  const slot = node as HTMLSlotElement;
+  const elements = slot.assignedElements() as HTMLElement[];
+  return { elements, parent: elements.at(0)?.parentElement };
+}
+
+/**
+ * Traverses and yields all focusable elements starting at `root`.
+ */
+function* getFocusableElements<T extends HTMLElement>(
+  root: HTMLElement | ShadowRoot,
+  cache?: WeakSet<HTMLElement>
+): Generator<T> {
+  if (!isDefined(globalThis.document)) {
+    return;
+  }
+
+  let node: T;
+  const _cache = cache ?? new WeakSet<HTMLElement>();
+
+  const visitor = document.createTreeWalker(
+    root,
+    NodeFilter.SHOW_ELEMENT,
+    (node) => shouldSkipElements(node, _cache)
+  );
+
+  // biome-ignore lint/suspicious/noAssignInExpressions: short-form
+  while ((node = visitor.nextNode() as T)) {
+    if (_cache.has(node)) {
+      continue;
+    }
+
+    if (node.shadowRoot) {
+      yield* getFocusableElements(node.shadowRoot, _cache);
```

Comment:
Hm, is it me or straight-up gathering focusable items will not work correctly with custom elements that employ `delegatesFocus`? In that case we should in theory strincly collect the first item or the one marked with `autofocus` if we're to respect the component's behavior? Probably not a scenario we're concerned with right now, just noting becuase this is done as generic solution.